world_positioning/src/worldP_ekf.py

```diff
@@ -0,0 +1,381 @@
+from worldP_listener import SubStateListener, ZedCamListener
+
+def quat_msg_to_array(q):
+    return np.array([q.w, q.x, q.y, q.z])
+
+def vec_to_msg(vec):
+    message = Vector3()
+    message.x = vec[0]
+    message.y = vec[1]
+    message.z = vec[2]
+    return message
+
+def quat_to_euler(q):
+    # From wikipedia (https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles)
+    # Copied from state estimation
+    w = q[0]
+    x = q[1]
+    y = q[2]
+    z = q[3]
+
+    euler = np.array([0,0,0])
+
+    # roll (x-axis rotation)
+    sinr_cosp = +2.0 * (w * x + y * z)
+    cosr_cosp = +1.0 - 2.0 * (x * x + y * y)
+    #angles[0] = np.arctan2(sinr_cosp, cosr_cosp)
+    
+
+    # yaw (z-axis rotation)
+    siny_cosp = +2.0 * (w * z + x * y)
+    cosy_cosp = +1.0 - 2.0 * (y * y + z * z)
+    #angles[2] = math.atan2(siny_cosp, cosy_cosp)
+
+    # pitch (y-axis rotation)
+    sinp = +2.0 * (w * y - z * x)
+    if np.abs(sinp) >= 1:
+        #angles[1] = np.sign(sinp) * math.pi # use 90 degrees if out of range
+        return np.array([ np.arctan2(sinr_cosp, cosr_cosp),
+                            np.sign(sinp) * math.pi,
+                            np.arctan2(siny_cosp, cosy_cosp)
+                            ])
+    else:
+        #angles[1] = math.asin(sinp)
+        return np.array([ np.arctan2(sinr_cosp, cosr_cosp),
+                            np.arcsin(sinp),
+                            np.arctan2(siny_cosp, cosy_cosp)
+                            ])
+
+    return euler
+
+def quat_msg_to_euler(q):
+    q_array = quat_msg_to_array(q)
+    return vec_to_msg(quat_to_euler(q_array))
+
+def quat_to_euler_jacobian(q):
+    return jacobian(quat_to_euler(q))
+
+class EKF:
+    def __init__(self):
+        # https://en.wikipedia.org/wiki/Extended_Kalman_filter
+        
+        ''' Model Description
+        x = state
+        u = inputs (sub position and orientation, object position in camera)
+        z = outputs (sub and object position)
+        P = uncertainty/variance matrix of state x
+
+        Dynamics: x[k+1] = f(x[k], u[k])
+        Outputs:    z[k] = h(x[k], u[k])
+
+        Linear Form:
+          x[k+1] = A*x[k] + B*u[k]
+            z[k] = C*x[k] + B*u[k]
+        '''
+
+        ''' Process Description
+
+        Each loop will do a prediction step based on the sub state,
+        to update the expected sub and world object positions and their variance.
+        
+        Then a measurement step, where it uses input from the zed camera's
+        detected objects to refine the prediction and varience
+        
+        Then there is a step where the new expected sub state and world state is converted
+        to a WorldState message which is then published.
+
+        '''
+
+        #first version: only 1 object in state
+
+        self.nState = 13 #number of elements in state
+
+        #n and m are only for 1 point
+        self.n = self.nState+3 #substate + world x,y,z of object
```

Comment:
yes, currently im running on the premise that its a single point for now. I'm not actually sure how we'd do multiple things, like can i change the size of the array as more data comes in or should i make enough array space for like 10 objects or smth. I'm not clear about what ros can do either, like whether it can output messages with variable size arrays as it is running. What do you think?

Reply:
Using one object may be good for initial testing. But we will need to be ready to switch to more objects. The easiest way to deal with multiple objects (I think) is just to have all of them in the state vector all the time. Luckily, we know what objects are part of the competition beforehand.

ROS can handle messages with variable-sized lists (and lists of messages). What we will also need to consider is how the vision system provides information to this ekf (ie: different messages/topics for each object, one message/topic with an id field, etc)